internal/database/txn/errors.go

```diff
@@ -48,3 +49,33 @@ func IsErrRetryable(err error) bool {
 
 	return false
 }
+
+func isConstraintError(err error) bool {
+	if err == nil || errors.Is(err, sqlite3.ErrLocked) || errors.Is(err, sqlite3.ErrBusy) {
+		return false
+	}
+	return isErrCode(err, sqlite3.ErrConstraintCheck) ||
+		isErrCode(err, sqlite3.ErrConstraintForeignKey) ||
+		isErrCode(err, sqlite3.ErrConstraintNotNull) ||
+		isErrCode(err, sqlite3.ErrConstraintPrimaryKey) ||
+		isErrCode(err, sqlite3.ErrConstraintTrigger) ||
+		isErrCode(err, sqlite3.ErrConstraintUnique) ||
+		isErrCode(err, sqlite3.ErrConstraintRowID)
+}
```

Comment:
This needs more thought, we should probably move this to somewhere better, that both the `internal/database` and `internal/database/txn` package can use. Potentially `internal/database/errors`.